Add linked record paths to `records.list_summaries`

mathemancer opened the pull request:
<!-- Concisely describe what the pull request does. -->

This adds a new parameter to the `records.list_summaries` RPC function, and new information to the response. This is to support a UI where one can list summaries from a table, along with checkboxes (or other boolean symbols) denoting whether or not a given summary is derived from a record which is linked via a simple many-to-many mapping table to a given record.

**Technical details**
<!-- Add any other information or technical details about the implementation; or delete the section entirely. -->

Example: List the summaries of the genres table, with a checkbox for whether the summary is for a genre record linked to the movie "Fellowship of the Ring", with ID 5.

There is a new parameter, `linked_record_path` in the `records.list_summaries` function, which takes a JSON of the following form:

```json5
{
    "record_pkey": 5, // record ID of "Fellowship of the Ring"
    "join_path": [
        [
            [
                17864, // movies OID
                1 // movies.id attnum
            ], [
                17808, // movies_genres OID
                2 // movies_genres.movie attnum
            ]
        ], [
            [
                17808, // movies_genres OID
                3 // movies_genres.genre attnum
            ], [
                17908, // genres OID
                1 // genres.id attnum
            ]
        ]
    ]
}
```

Some gotchas:

- The join path must always have a length of 2, since it's denoting a join through a simple many-to-many mapping table in this context.
- The join path is in the order from the `movies` table _to_ the `genres` table, when listing summaries from `genres`. This is to support the UI, and also aligns with the fact that the front end will have such a join path available since they'll have the response to `tables.list_joinable` in hand, but called on the `movies` (rather than `genres`) table.

The response adds a new top-level key, `mapping`, resulting in a JSON response like this:

```json
{
  "count": 100,
  "results": [
    {"key": 3, "summary": "Fantasy"},
    {"key": 34, "summary": "Sci-Fi"},
    {"key": 45, "summary": "Action"}
  ],
  "mapping": {
    "join_table": 17808,
    "joined_values": {
      "3": 4734,
      "45": 5467
    }
  }
}
```

The `join_table` in this example would be `movies_genres`. The keys of the `joined_values` object are the keys associated with each record summary in the `genres` table, i.e., they're primary key values of the `genres` table. The values of the `joined_values` table are primary keys of the `movies_genres` table, to support delinking a genre from a movie by deleting the associated record from the `movies_genres` table if desired, and also to allow the front end to show a checked box next to a summary. I.e., if there exists a key, value pair in `joined_values` for a given key in the returned record summaries, then there is a mapping to the specified record in the request, and so the box should be "checked".

## Checklist
<!-- Replace  the [ ] with [x] to check the boxes. --> 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [ ] My pull request targets the `develop` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

mathemancer committed:
add SQL function to build mapping table values

mathemancer committed:
add new functions to all objects table

mathemancer committed:
wire linked record paths up to RPC function

mathemancer committed:
improve documentation of new functionality

seancolsen commented:
@mathemancer can you re-stack this work on top of `cross-table-mvp` instead of `develop`? (I'm working on top of this now.)

seancolsen commented:
By the way, these API changes are working well for me! 😄

mathemancer commented:
@seancolsen done, but we should merge develop into `cross-table-mvp`. Do you mind if I do? Not sure what your current status is on that branch.

mathemancer committed:
modify to work with older Postgres versions

mathemancer committed:
update SQL tests to match new return format

mathemancer committed:
remove unused private method

mathemancer committed:
tidy up documentation of new logic and behavior

mathemancer committed:
add tests to make sure pagination and map work

mathemancer commented:
@seancolsen ~Actually, so that this is reviewable, I need to merge develop into the underlying branch. I hope that doesn't cause problems. If it does, I guess we can figure out what to do from there.~

Apparently the MVP branch is branched from master or something? Not sure. Regardless, I reverted my merging of develop into that branch, and changed the target of this PR to develop to make it easier to review. If your intent was for me to rebase off of that branch or something, I'd need to do more spelunking or otherwise figure out what that branch is growing off of.

mathemancer committed:
change to arrays for mapping pkeys

mathemancer committed:
update docs to match new response structure

seancolsen commented:
BTW, I have #4999 on top of this now.

seancolsen commented:
@Anish9901 I assigned you to review this so that we can get it merged now that I have another PR stacked on it.

Anish9901 commented on db/sql/05_msar.sql:
```diff
@@ -5765,12 +5765,54 @@ END;
+$$ LANGUAGE SQL;
+
+
+CREATE OR REPLACE FUNCTION msar.get_simple_mapping_join_cte(
```
We can make this into a SQL function instead of plpgsql, as we have `msar.raise_exception` to raise exceptions within SQL functions. But, I think it's fine for now.

Anish9901 reviewed (approved):
Looks great! Pointed out a minor nitpick, but we can handle it in another PR.
